### PR TITLE
Get user by auth0 token

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

The PR adds API route `/api/secure/user/fromtoken` to obtain users by Auth0 token, and removes route `/api/secure/{entity}/{find}/{value}` to restrict what can be queried.

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Doulble 'and'

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Typo 'overriden' should be 'overridden'

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: I'm wondering if all actions related to user auth should be outside of the controllers as similar checks are likely to be done by other controllers. For instance, under the monitored-trip branch I have created: 

https://github.com/ibi-group/otp-middleware/blob/monitored-trip/src/main/java/org/opentripplanner/middleware/auth/Auth0Utils.java

**Review by br648 — COMMENTED** [review]
I'm assuming once this is approved a similar update will be required to the monitor trip controller?

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (landonreed) [reply src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: So that everyone is aware, [`Auth0Connection#checkUser`](https://github.com/ibi-group/otp-middleware/blob/94dcefa65446530ba5854afb442aac7d416c5312/src/main/java/org/opentripplanner/middleware/auth/Auth0Connection.java#L38) checks that the user is authorized. This check is run on every request method [behind a `api/secure/*` or `api/admin/*` route](https://github.com/ibi-group/otp-middleware/blob/fced81463189960895df9cfa4eee26eef9637151/src/main/java/org/opentripplanner/middleware/spark/Main.java#L79-L86).

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: This else condition should never happen on an endpoint in this controller (see https://github.com/ibi-group/otp-middleware/pull/22/files#r432599004).

File: src/test/java/org/opentripplanner/middleware/controllers/api/UserControllerTest.java
Comment (landonreed) [inline src/test/java/org/opentripplanner/middleware/controllers/api/UserControllerTest.java]: These tests feel a little too contrived. If we're making controller tests, I would rather just send http requests to the API to verify the JSON response.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Needs some comments.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (landonreed) [reply src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: This javadoc feels a little sparse. How about something like:

```
This method implements a set of basic HTTP Spark methods (e.g., getOne, getMany, delete) for CRUD operations. It can be overridden by child classes to supplement additional methods through the ApiEndpoint (after supplemental methods are added, be sure to call the super method to add the basic methods).
```

And now that I'm taking a closer look: should `apiEndpoint` just be a field on this class? And I don't think the `makeEndpoint` method needs to return the ApiEndpoint, right?

**Review by landonreed — CHANGES_REQUESTED** [review]
Looks great, I do have a few suggestions though:
- changes for `makeEndpoint` signature/javadoc
- could we make the controller tests handle actual http requests to the user controller api endpoint?

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Good catch, it is no longer necessary.
I had the return because I initially thought doing a chain add, e.g.
```java
return super.makeEndPoint()
.get(path(...))
.options(path(...))
```
but we have to call `super` after adding the custom routes. I'm updating the comments in this method accordingly.

Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: It don't think `apiEndpoint` can become a field because the parent, `baseEndPoint`, is not available until spark-java calls the `bind` method.

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Hmm good point. We should consider making that profile a field?

File: src/test/java/org/opentripplanner/middleware/controllers/api/UserControllerTest.java
Comment (binh-dam-ibigroup) [reply src/test/java/org/opentripplanner/middleware/controllers/api/UserControllerTest.java]: This unit test is an additional confidence check before going all in at the HTTP request level.

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Auth0Connection#checkUser checks that the user is authorized prior to this endpoint being reached. Therefore the profile parameter already contains the OtpUser (or not) so this line is not required. The OtpUser can be obtained from profile.otpUser.

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Brackets.

**Review by br648 — CHANGES_REQUESTED** [review]
http://localhost:4567/api/secure/user/ returns a 404 "No API route configured for this path". My understanding is that this should be implemented as part of the ApiController.buildEndPoint call. It is listed under the Swagger UI endpoint (http://localhost:4567/) so I'm not sure why it fails.

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Thanks for pointing this one out, @br648, that simplifies *a lot*.

**Comment by binh-dam-ibigroup** [comment]
> http://localhost:4567/api/secure/user/ returns a 404 "No API route configured for this path". My understanding is that this should be implemented as part of the ApiController.buildEndPoint call. It is listed under the Swagger UI endpoint (http://localhost:4567/) so I'm not sure why it fails.

@br648, it is either `http://localhost:4567/api/secure/user` or `http://localhost:4567/api/secure/user/{mongo_id}`.

**Comment by binh-dam-ibigroup** [comment]
> * could we make the controller tests handle actual http requests to the user controller api endpoint?
@landonreed after simplification, that test file has been removed for the scope of the changes of this PR.

**Comment by br648** [comment]
> 
> 
> > http://localhost:4567/api/secure/user/ returns a 404 "No API route configured for this path". My understanding is that this should be implemented as part of the ApiController.buildEndPoint call. It is listed under the Swagger UI endpoint (http://localhost:4567/) so I'm not sure why it fails.
> 
> @br648, it is either `http://localhost:4567/api/secure/user` or `http://localhost:4567/api/secure/user/{mongo_id}`.

Sorry, my mistake.

`http://localhost:4567/api/secure/user` does return all users, but this will be addressed in PR https://github.com/ibi-group/otp-middleware/pull/21

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: This should be replaced with Auth0Connection.getUserFromRequest(Request req);

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Can be deleted.

**Review by br648 — CHANGES_REQUESTED** [review]
It is still possible to retrieve a user that doesn't match the authorized user:

http://localhost:4567/api/secure/user/841c1d8e-9ae1-4fef-9156-27cf77113a73-a-diff-user

I've addressed this issue for 'getMany' in https://github.com/ibi-group/otp-middleware/pull/21 I'm now wondering if I should make a similar update to 'getOne' and 'deleteOne'? @landonreed @binh-dam-ibigroup thoughts?

**Comment by binh-dam-ibigroup** [comment]
> It is still possible to retrieve a user that doesn't match the authorized user:
> 
> http://localhost:4567/api/secure/user/841c1d8e-9ae1-4fef-9156-27cf77113a73-a-diff-user
> 
> I've addressed this issue for 'getMany' in #21 I'm now wondering if I should make a similar update to 'getOne' and 'deleteOne'? @landonreed @binh-dam-ibigroup thoughts?

@br648 How about the `canBeManagedBy` method in class `AbstractUser`? https://github.com/ibi-group/otp-middleware/blob/4b0cb1ed711fb7ec834563340d2eafb61abba4d0/src/main/java/org/opentripplanner/middleware/models/AbstractUser.java#L39

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Good suggestion.

**Comment by binh-dam-ibigroup** [comment]
@br648 `deleteOne` already has the check for whether a user can delete another user, so you could add the same check to `getOne`: https://github.com/ibi-group/otp-middleware/blob/4b0cb1ed711fb7ec834563340d2eafb61abba4d0/src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java#L152

**Comment by br648** [comment]
> 
> 
> > It is still possible to retrieve a user that doesn't match the authorized user:
> > http://localhost:4567/api/secure/user/841c1d8e-9ae1-4fef-9156-27cf77113a73-a-diff-user
> > I've addressed this issue for 'getMany' in #21 I'm now wondering if I should make a similar update to 'getOne' and 'deleteOne'? @landonreed @binh-dam-ibigroup thoughts?
> 
> @br648 How about the `canBeManagedBy` method in class `AbstractUser`?
> 
> https://github.com/ibi-group/otp-middleware/blob/4b0cb1ed711fb7ec834563340d2eafb61abba4d0/src/main/java/org/opentripplanner/middleware/models/AbstractUser.java#L39

@binh-dam-ibigroup yes I think that will work.

**Comment by binh-dam-ibigroup** [comment]
Latest changes pushed, @landonreed, @br648, @evansiroky.

**Comment by br648** [comment]
> 
> 
> @br648 `deleteOne` already has the check for whether a user can delete another user, so you could add the same check to `getOne`:
> 
> https://github.com/ibi-group/otp-middleware/blob/4b0cb1ed711fb7ec834563340d2eafb61abba4d0/src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java#L152

@binh-dam-ibigroup thanks, yes will do that.

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: typo: applicable applicable

File: src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java]: Can this be formatted better so that the function arguments have the same offset from the start of the line?

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (evansiroky) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Please have better formatting here with spaces after commas. Also, the error string doesn't seem right because there will be an Auth0 user, just not an otpUser.

**Review by evansiroky — DISMISSED** [review]
Just three little comments

**Review by landonreed — APPROVED** [review]
LGTM, once other comments are addressed.

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: @evansiroky, A valid Auth0 user id does not guarantee existence of a corresponding `OtpUser` entity with the proper `auth0UserId` attribute, and the error case covers that. Example: First-time users who just signed up on Auth0 and have not entered their preferences.

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Does this need to consider an apiUser and adminUser users?

**Review by br648 — COMMENTED** [review]
Looks good, just the one comment. Once that has been address I can approve.

File: src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: `UserController` is for `OtpUser`s.

Comment (br648) [reply src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java]: Of course. Thanks for clarifying.

**Review by br648 — APPROVED** [review]
All good.

File: src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java]: Just noticed that this log statement doesn't make use of the `ROOT_ROUTE` + `TOKEN_PATH`.

File: src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java
Comment (landonreed) [inline src/main/java/org/opentripplanner/middleware/controllers/api/OtpUserController.java]: This description looks to be incorrect.

**Comment by binh-dam-ibigroup** [comment]
Latest changes noted by @evansiroky  and @landonreed have been incorporated. I'm proceeding to merge.

**Pull request merged**
